### PR TITLE
[SPARK-15632][SQL]Typed Filter should NOT change the Dataset schema

### PR DESCRIPTION
**Pull request by clockfly**
## What changes were proposed in this pull request?

This PR makes sure the typed Filter doesn't change the Dataset schema.

**Before the change:**

```
scala> val df = spark.range(0,9)
scala> df.schema
res12: org.apache.spark.sql.types.StructType = StructType(StructField(id,LongType,false))
scala> val afterFilter = df.filter(_=>true)
scala> afterFilter.schema   // !!! schema is CHANGED!!! Column name is changed from id to value, nullable is changed from false to true.
res13: org.apache.spark.sql.types.StructType = StructType(StructField(value,LongType,true))

```

SerializeFromObject and DeserializeToObject are inserted to wrap the Filter, and these two can possibly change the schema of Dataset.

**After the change:**

```
scala> afterFilter.schema   // schema is NOT changed.
res47: org.apache.spark.sql.types.StructType = StructType(StructField(id,LongType,false))
```
## How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 1]
```diff
@@ -225,6 +225,12 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       "b")
   }
 
+  test("filter should not change the underlying logical plan schema") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: add the jira ticket number here

Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: and maybe "typed filter should preserve the underlying logical schema" ?

**Comment by cloud-fan** [comment]
we can also remove the `EmbedSerializerInFilter` and its related tests

**Comment by rxin** [comment]
maybe ignore them for now? useful as reference when we add those optimizations.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -1961,11 +1961,11 @@ class Dataset[T] private[sql](
    */
   @Experimental
   def filter(func: FilterFunction[T]): Dataset[T] = {
-    val deserialized = CatalystSerde.deserialize[T](logicalPlan)
+    val deserializer = UnresolvedDeserializer(encoderFor[T].deserializer)
     val function = Literal.create(func, ObjectType(classOf[FilterFunction[T]]))
-    val condition = Invoke(function, "call", BooleanType, deserialized.output)
-    val filter = Filter(condition, deserialized)
-    withTypedPlan(CatalystSerde.serialize[T](filter))
+    val condition = Invoke(function, "call", BooleanType, Nil)
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Why the arguments is `Nil`?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Probably also add a test case for the Java API in `JavaDatasetSuite`?

**Comment by liancheng** [comment]
@rxin By "ignore" did you mean remove the rule from the batch but keep the code lines of the rule?

**Comment by rxin** [comment]
I was referring to the tests.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TypedFilterOptimizationSuite.scala [hunk 3]
```diff
@@ -57,7 +57,9 @@ class TypedFilterOptimizationSuite extends PlanTest {
     comparePlans(optimized, expected)
   }
 
-  test("embed deserializer in filter condition if there is only one filter") {
+  // TODO: Remove this after we completely fixes SPARK-15632 by adding optimization rules
+  // For typed filters.
```
Comment (liancheng) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TypedFilterOptimizationSuite.scala]: Nit:

- fixes => fix
- For => for

**Comment by liancheng** [comment]
LGTM except for a minor comment, and pending Jenkins.

**Comment by liancheng** [comment]
@cloud-fan I think we can also disable `EmbedSerializerInFilter` now?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TypedFilterOptimizationSuite.scala (hunk 3 above)
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/TypedFilterOptimizationSuite.scala]: we want to "reenable this", not "remove this" right?

**Comment by cloud-fan** [comment]
I think we should just remove `EmbedSerializerInFilter`, as we need a totally different optimization strategy for the new typed filter.

**Comment by clockfly** [comment]
@liancheng @cloud-fan 

About EmbedSerializerInFilter, I have noted in source code:
`TODO: Remove this after we completely fixes SPARK-15632 by adding optimization rules For typed filters.`

We can remove EmbedSerializerInFilter in follow up PR when we finished the optimizers for typed filter.

**Comment by liancheng** [comment]
Merging to master and branch-2.0. Thanks!

**Pull request closed without merging**